enhancements/mco-kargs-day1-support.md

```diff
@@ -0,0 +1,347 @@
+---
+title: Kernel Arguments - Day 1 Support
+authors:
+  - "@ericavonb"
+reviewers:
+  - "@cgwalters"
+  - "@crawford"
+  - "@imcleod"
+  - "@runcom"
+approvers:
+  - "@cgwalters"
+  - "@crawford"
+  - "@imcleod"
+  - "@runcom"
+
+creation-date: 2019-09-12
+last-updated: 2019-09-23
+status: **provisional**|implementable|implemented|deferred|rejected|withdrawn|replaced
+---
+
+# Kernel Args - Day 1 Support
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs]
+
+## Summary
+
+This enhancement proposes adding the ability for admins to configure custom
+kernel arguments during initial RHCOS node configuration. With this enhancement,
+both OS upgrade to the latest [machine-os-content](https://github.com/openshift/machine-config-operator/blob/master/docs/OSUpgrades.md#os-updates)
+and custom kernel args can be applied to a node during the first boot of the
+machine, followed by a single reboot.
+
+## Motivation
+
+Kernel arguments for RHCOS can be set today using a MachineConfig as a
+[`day-2` operation](https://github.com/openshift/machine-config-operator/blob/master/docs/MachineConfiguration.md#kernelarguments),
+i.e. after cluster install, incurring an additional reboot to apply. Even if the
```

Comment:
Is the uncoordinated reboot at install-time that important? No major components will be scheduled on the nodes that early in node's bootstrap process..?

Reply:
It's particularly bad for the control plane - the etcd cluster will form but then each MCD will try draining its node; we'll usually block on etcd-quorum-guard, but...  Similarly for workers we will end up fighting the PDB for the router, etc.

Reply:
I am under the assumption that the machine boots, downloads the machine-os-content reboots, sets the kernel args using rpm-ostree and reboots.
^^ in this scenario we would see both the reboots before etcd membership is in question.

But if the flow is different maybe i'm missing something here and I would like this doc to clarify the current process in detail?

Reply:
Today is goes:
boot a - bootstrap image
* machine boots up using bootstrap image specified in the installer
* machine fetches ignition payload from bootstrap machine
* services start up container runtime, MCD pivot from binary on host
* pivot pulls machine-os-content, reboots

Boot b - latest rhcos
* machine reboots with up-to-date rhcos image
* services start up: runtime, pivot binary from latest rhcos, pivot sees its at latest, no action needed
* kubelet starts up, starts static pods/etcd
* kube control plane spun up
* MCO, other cluster components land on a node
* MCO spins up MCC, MCS, MCDs
* MCC spins up etcd-quorum-guard daemonset
* MCC starts template controller which renders MCs
* MCD lands on node, pulls its MC, sees it has k-args to apply (plus whatever else), sets them using rpm-ostree, waits on drain to reboot

There’s a race condition between each master node’s etcd joining the cluster, quorum-guard containers starting up, MCD starting up and then doing the drain and reboot, and probably other places as well.

We’re proposing:

boot a - bootstrap image
* machine boots up using bootstrap image specified in the installer
* machine fetches ignition payload from bootstrap machine
* services start up container runtime, MCD from binary on host
* MCD uses MC on disk, pulls machine-os-content image specified in MC, applies the k-args, reboots

boot b
* same as above but MCD sees machine is already up to date and so doesn’t reboot
This allows the cluster to complete openshift installation before applying changes (from user once they can access cluster, or an update coordinated by the CVO) that require reboots